### PR TITLE
[sclang] Add Ableton Link support (rebased)

### PR DESCRIPTION
**Pull request by snappizz**
Purpose and Motivation
----------------------

Add Ableton Link support. Rebase of #3351 to fix conflicts and restart 100+ comment discussion from a blank slate.

Types of changes
----------------

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by gilfuser** [comment]
Hi there.
I know this is something for the future, but I gave it a try, to see if it would work. I'm posting the results as I think it may help to have it going on.
I changed all the modified files and created the necessary ones to build SC already with this PR.
The build failed because of this:
```
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp: In member function ‘virtual double LinkClock::BeatsToSecs(double) const’:
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:42:25: error: ‘const class ableton::Link’ has no member named ‘captureAppTimeline’; did you mean ‘captureAppSessionState’?
   auto timeline = mLink.captureAppTimeline();
                         ^~~~~~~~~~~~~~~~~~
                         captureAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp: In member function ‘virtual double LinkClock::SecsToBeats(double) const’:
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:48:25: error: ‘const class ableton::Link’ has no member named ‘captureAppTimeline’; did you mean ‘captureAppSessionState’?
   auto timeline = mLink.captureAppTimeline();
                         ^~~~~~~~~~~~~~~~~~
                         captureAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp: In lambda function:
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:75:25: error: ‘class ableton::Link’ has no member named ‘captureAppTimeline’; did you mean ‘captureAppSessionState’?
   auto timeline = mLink.captureAppTimeline();
                         ^~~~~~~~~~~~~~~~~~
                         captureAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:77:9: error: ‘class ableton::Link’ has no member named ‘commitAppTimeline’; did you mean ‘commitAppSessionState’?
   mLink.commitAppTimeline(timeline);
         ^~~~~~~~~~~~~~~~~
         commitAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp: In constructor ‘LinkClock::LinkClock(VMGlobals*, PyrObject*, double, double, double)’:
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:101:24: error: ‘class ableton::Link’ has no member named ‘captureAppTimeline’; did you mean ‘captureAppSessionState’?
  auto timeline = mLink.captureAppTimeline();
                        ^~~~~~~~~~~~~~~~~~
                        captureAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:104:8: error: ‘class ableton::Link’ has no member named ‘commitAppTimeline’; did you mean ‘commitAppSessionState’?
  mLink.commitAppTimeline(timeline);
        ^~~~~~~~~~~~~~~~~
        commitAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp: In member function ‘void LinkClock::SetAll(double, double, double)’:
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:110:24: error: ‘class ableton::Link’ has no member named ‘captureAppTimeline’; did you mean ‘captureAppSessionState’?
  auto timeline = mLink.captureAppTimeline();
                        ^~~~~~~~~~~~~~~~~~
                        captureAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:118:8: error: ‘class ableton::Link’ has no member named ‘commitAppTimeline’; did you mean ‘commitAppSessionState’?
  mLink.commitAppTimeline(timeline);
        ^~~~~~~~~~~~~~~~~
        commitAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp: In member function ‘void LinkClock::SetTempoAtBeat(double, double)’:
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:124:24: error: ‘class ableton::Link’ has no member named ‘captureAppTimeline’; did you mean ‘captureAppSessionState’?
  auto timeline = mLink.captureAppTimeline();
                        ^~~~~~~~~~~~~~~~~~
                        captureAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:131:8: error: ‘class ableton::Link’ has no member named ‘commitAppTimeline’; did you mean ‘commitAppSessionState’?
  mLink.commitAppTimeline(timeline);
        ^~~~~~~~~~~~~~~~~
        commitAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp: In member function ‘void LinkClock::SetTempoAtTime(double, double)’:
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:136:24: error: ‘class ableton::Link’ has no member named ‘captureAppTimeline’; did you mean ‘captureAppSessionState’?
  auto timeline = mLink.captureAppTimeline();
                        ^~~~~~~~~~~~~~~~~~
                        captureAppSessionState
/home/gilfuser/builds/supercollider/lang/LangPrimSource/SC_Ableton_Link.cpp:142:8: error: ‘class ableton::Link’ has no member named ‘commitAppTimeline’; did you mean ‘commitAppSessionState’?
  mLink.commitAppTimeline(timeline);
        ^~~~~~~~~~~~~~~~~
```

**Comment by jamshark70** [comment]
External sync popped up again on my radar, so I thought I would give this a try.

I get a cmake error.

```
CMake Error in lang/CMakeLists.txt:
  Imported target "Ableton::Link" includes non-existent path

    "/home/dlm/share/supercollider/external_libraries/link/cmake_include/../modules/asio-standalone/asio/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.
```

How to resolve?

@gilfuser -- it looks like you might need to monkey around with the git submodules by hand. The PR adds the following to supercollider/.gitmodules:

```
[submodule "external_libraries/link"]
	path = external_libraries/link
	url = https://github.com/Ableton/link
```

But this is not automatically added to supercollider/.git/config, so git then does not check out the module.

So I added this to .git/config:

```
[submodule "external_libraries/link"]
	url = https://github.com/Ableton/link
```

Then `git submodule update` properly fetched the module.

**Comment by SimonZimmer** [comment]
Hey @jamshark70,

Just checking if you got it working with your .git/config fix? 
Or are there still some issues?
Very excited for this pull request btw!

best,
Simon

**Comment by brianlheim** [comment]
@jamshark70 it sounds like you should have originally run `git submodule update --init --recrusive`. the link repo has nested submodules itself, and `update` on its own won't check out new submodules, that's what `--init` is for.

**Comment by jamshark70** [comment]
Oh... now I remember. The earlier discussion had considered an `aheadOfTime` parameter... seems to have been overlooked in this iteration.

I feel strongly that we need to revive this idea. Users will think: "If I've got Ableton Live running, and I start a pattern on a LinkClock with `quant: -1`, then I expect to *hear* the first SC event in sync with Live's downbeat."

As far as I can see (please correct me if I'm wrong), as it stands now, when Live's clock is on the downbeat, SC's clock will also be on the downbeat. Then SC will send a server message with latency, and you will hear the event *out of sync* with Live, and users will think "Well this doesn't work at all."

If I recall, the proposal was that `LinkClock` would have an `aheadOfTime` property, in seconds. If you set `myLinkClock.aheadOfTime = s.latency`, then you would hear good sync.

To do that, LinkClock's `beats` within sclang would be "link time" + (aheadOfTime * tempo), but it would use "link time" when communicating externally. I.e., Live says "it's beat 12" and SC would say externally "it's beat 12," but if sclang asks for the clock's beats at that moment, it would get 12.2 (assuming tempo = 1 and latency = 0.2). And, when you schedule that pattern with `quant: -1`, it would happen at sclang time 12.0, but "link time" 11.8 (and then SC's messaging latency would take care of the rest). (There are some patterns that depend on `beatInBar` -- so SC's should appear, within sclang, to schedule quant:1 on an integer beat, while positioning that moment in time earlier than the external timebase.)

Link's readme says: "... the audio system's output latency should be added to system time values before passing them to Link methods." We have additional latency to consider -- if we don't factor in our messaging latency, then we aren't using Link correctly.

File: lang/LangPrimSource/SC_Ableton_Link.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	if(err) throw err;
+
+	mLink.enable(true);
+	mLink.setTempoCallback([this](double bpm) {
+		double secs = elapsedTime();
+		double tempo = bpm / 60.;
+
+		auto timeline = mLink.captureAppTimeline();
+		double beats = timeline.beatAtTime(hrToLinkTime(secs), mQuantum);
+		mLink.commitAppTimeline(timeline);
```
Comment (colinsullivan) [inline lang/LangPrimSource/SC_Ableton_Link.cpp]: Why `commitAppTimeline` here ? Isn't this method meant to be used to push changes back to Link ? Seems to me at this point no changes need to be committed, the LinkClock is just reading the state from Link.

**Comment by colinsullivan** [comment]
While I do not want to detract from the importance of the other points in this conversation, it is worth noting this submodule is at a commit from 2017.  The Link API has changed since, including some methods used in this PR (`Timeline` is now `SessionState`).

File: lang/CMakeLists.txt [hunk 2]
```diff
@@ -222,6 +223,14 @@ if (NOT WIN32)
 		PROPERTY OUTPUT_NAME sclang)
 endif()
 
+if (SC_ABLETON)
```
Comment (dvzrv) [inline lang/CMakeLists.txt]: Could we potentially check for the library being installed on the system and use that?

**Comment by mzero** [comment]
Side note: When this finally lands, the build instructions in the wiki will need to be updated to have the git submodules be init'd and update'd recursively:

    git update --init --recursive

should do it. 

This operation appears on the building for RaspberryPi page, and the Cheat Sheet for git page, and perhaps others.

File: SCClassLibrary/External/Ableton/LinkClock.sc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	prSetQuantum { |quantum|
+		_LinkClock_SetQuantum;
+		^this.primitiveFailed
+	}
+
+	// run tempo changed callback
+	prTempoChanged { |tempo, beats, secs, clock|
+		tempoChanged.value(tempo, beats, secs, clock);
```
Comment (mzero) [inline SCClassLibrary/External/Ableton/LinkClock.sc]: wondering if this should be:

```suggestion
		tempoChanged.value(tempo, beats, secs, clock);
		this.changed(\tempo);
```

Other changes to the tempo in `TempoClock` send that notification. If `LinkClock` had, then my performance patch using `TempoClock` would have been able to be Link enabled by just changing the class of its clock object!

Comment (jamshark70) [reply SCClassLibrary/External/Ableton/LinkClock.sc]: Yes, `changed(\tempo)` should be mandatory.

Also, if the link quantum changes, then we need to update beatsPerBar and also do `this.changed(\meter)`.

The meter change could be tricky because of `baseBarBeats`.

**Comment by jamshark70** [comment]
I have a proposal to handle the latency compensation problem.

https://github.com/jamshark70/supercollider/commit/0cbb3f6c263e8956faf3066bed137878498d4006

Not sure about:

- C++ style problems?
- I tried this only today linking sclang <--> sclang. Haven't had a chance to try sclang <--> Ableton Live. There's a chance it might not pan out. But I think my logic is sound.

Overview:

- Add a member variable `mLatency` to the C++ LinkClock class. (I'm using a member variable because we need it in `double BeatsToSecs()` and `double SecsToBeats()`, which appear to be crucial to TempoClock's internal sync and as such might be called very often. I think it's a bad idea to incur the cost of sclang slot access in potentially high-traffic functions.)

- In this version, the LinkClock's default latency is 0. I thought about making the default `Server.default.latency`, but it's possible to have multiple server objects with different latencies, used with the same clock. Sclang can't guess the user's relationship between clock and server -- rather than make some silly rule, I'm initializing to 0 and the user should override as desired. (Documentation issue -- I haven't written help for this, not until I'm sure it's working.)

- Simple case: tempo = 1 bps, for argument let's assume beats == seconds. I'm assuming, if say Live is going to play a sound at 12.0 seconds (shared timebase), and SC's latency is 0.2 sec, then SC's clock needs to read 12.0 beats at 11.8 seconds. So `BeatsToSecs()` should convert, then subtract latency. Likewise, `SecsToBeats()` should add latency, then convert.

To test, I ran two sclang instances. In both of them, server latency and clock latency match, but this is different between the two instances. This simulates the idea that Live's latency compensation is different from SC's, but the sounding result should line up.

```supercollider
// sclang #1
s.options.maxLogins = 2;
s.boot;

// LinkClock latency should match server
t = LinkClock(1).latency_(0.25); s.latency = 0.25;

p = Pbind(\degree, 0, \dur, 1, \legato, 0.1, \pan, -1, \time, Pfunc { thisThread.beats }).trace.play(t, quant: 1);
```

```supercollider
// sclang #2
Server.default = s = Server.remote(\remote, NetAddr("127.0.0.1", 57110));

t = LinkClock(1).latency_(0.1); s.latency = 0.1;

p = Pbind(\degree, 7, \dur, 1, \legato, 0.1, \pan, 1, \time, Pfunc { thisThread.beats }).trace.play(t, quant: 1);
```

And the result is...

- The trace in the two windows shows events being generated at different times (so the clocks, with different latency compensation values, adjust their timing differently -- as expected).

- The `\time` in the events, in beats, conforms to `quant: 1` (avoiding problems with patterns like PfinQuant that use `nextTimeOnGrid` internally).

- Left and right channels sound in sync.

I'll try to track down someone who uses Live and make sure it works with a non-SC peer, but I think my logic is right and this test shows that two clients with different latency compensation sound in sync.

I tried changing tempo also. There's still a problem with that -- speeding up the tempo may cause some scheduled functions to be skipped, and unfortunately I don't see this happening with latency == 0. Probably a lower latency value would make this less likely.

I think this is a step in the right direction, though. Comments?

**Comment by jamshark70** [comment]
> I tried changing tempo also. There's still a problem with that -- speeding up the tempo may cause some scheduled functions to be skipped, and unfortunately I don't see this happening with latency == 0.

After some more poking around, I found a discrepancy that might explain that issue.

https://github.com/jamshark70/supercollider/blob/0cbb3f6c263e8956faf3066bed137878498d4006/lang/LangPrimSource/SC_Ableton_Link.cpp#L129

`BeatsToSecs()` applies `mLatency` to the seconds value calculated from beats and tempo.

But, in my first attempt, `LinkClock::SetTempoAtBeat()` does not. It seems possible to me that this puts the clock into an improper state, causing seconds to be calculated incorrectly. (I have testing data showing that, if the tempo increases at the "wrong" time, beats2 > beats1 may correspond to seconds2 < seconds1, causing events to evaluate out of order).

Which leads to my C++ question: `timeline.timeAtBeat(...)` is some sort of std::chrono::duration type, and I can't simply subtract a `double` number of seconds.

How do I cast `mLatency` into a `std::chrono` type?

(I looked on stack exchange and I can't make any sense out of it.)

Another possible issue is https://github.com/jamshark70/supercollider/blob/0cbb3f6c263e8956faf3066bed137878498d4006/lang/LangPrimSource/SC_Ableton_Link.cpp#L80, which doesn't add mLatency to secs as is done elsewhere. I tested that but it doesn't fix the problem.

**Comment by mzero** [comment]
This approach, a latency feature of `LinkClock` that is measured in seconds, and applies negatively to the scheduling (that is, positive latency makes the beats be early), is exactly what I was after in the Slack channel.

Not to complicate things - but wouldn't `TempoClock` benefit from the feature as well? Perhaps it is best to do it first here - and if it proves desirable, generalize it later.

I'm using this with Live and have a set up for testing. I'd be happy to test your proposed patch if you send it to me... or I can hand re-create it.

**Comment by jamshark70** [comment]
@mzero you should be able to check out my branch. Borrowing github's "merge at command line" instructions:

```
git checkout -b hjh-topic/hjhLinkLatency develop
git pull git://github.com/jamshark70/supercollider.git topic/hjhLinkLatency
```

(I think...?)

**Comment by jamshark70** [comment]
And, thinking further, now I remember why I had argued months ago that this case was always going to be problematic.

The assumption behind latency compensation is that beat 10.0 is the time point when the sound should hit the speakers, and it should be the same time point for all peers.

In my tempo-change bug reproducer, I had:

- Peer 1: Tempo = 60, for argument's sake assume initially that beats == seconds, latency 250 ms.
- Peer 2: Tempo = 60, assume initially that beats == seconds, latency 100 ms.

Peer 1's Pbind, above, puts note-on on integer beats and note-off 0.1 beats later. To compensate for latency, in seconds it's note-on = x.75, note-off = x.85.

Peer 2 schedules a tempo change for e.g. 9.88 beats.

I had been thinking that this corresponds to seconds = 9.78, and this should be the stable, shared time reference. But that's wrong. A tempo change at beat z should change the tempo at beat z in all peers.

The proper way to think of the tempo change scheduling is that it's 0.12 beats before the note-on beat -- so, at the initial tempo of 1, P1 needs to process the tempo change 0.12 seconds before it processes its note-on: 9.75 sec (process note-on) - (0.12 beats * 1 sec/beat) = 9.63 sec. But P2 hasn't even generated the tempo change yet -- that's coming 0.15 beats later because of the difference in latencies.

So it's literally impossible, currently, for P1 to handle the tempo change correctly.

I think the only robust solution is if Link has a way to specify a future tempo change, which peers can query and prepare for. I'm pretty sure Link doesn't have such a feature.

There may not be much we can do except document that very large tempo changes are risky.

But... the most common use case for Link integration will be dance or ambient sets, where the tempo and meter are likely to be stable. I don't think we should withhold the feature because an unusual case will be difficult.

**Comment by jamshark70** [comment]
https://github.com/jamshark70/supercollider/tree/topic/hjhLinkLatency

- Latency compensation
- Minor fixes
  - Use `errWrongType` instead of `errFailed` where applicable
  - Wake up the clock upon a tempo change
  - Remove an unnecessary timeline commit
- Update Link to v3.0.2 (latest tagged version from Ableton)
- Notify dependents when the tempo changes externally

I believe this addresses most of the issues noted above.

I noted in the dev conference call minutes a couple of weeks ago that this feature was considered "stalled." Maybe we can un-stall it?

File: lang/LangPrimSource/SC_Ableton_Link.cpp (hunk 1 above)
Comment (jamshark70) [reply lang/LangPrimSource/SC_Ableton_Link.cpp]: You're right. My proposed fixes remove it: https://github.com/jamshark70/supercollider/commit/f9f38e791881f1568f364b1c9f3ad88922cac4dd

Tested and there is no ill effect.

File: SCClassLibrary/External/Ableton/LinkClock.sc (hunk 3 above)
Comment (jamshark70) [reply SCClassLibrary/External/Ableton/LinkClock.sc]: Fix proposed: https://github.com/jamshark70/supercollider/commit/a636e7b886d7ecc81f5121ef0239ed58094cd0f5

Also, Link (seemingly by design) does not propagate quantum changes. So we're off the hook for the meter-change question.

**Comment by jamshark70** [comment]
> Workaround for the tempo change case above.

https://github.com/jamshark70/supercollider/commit/8bc18022624e6b913395ef11c954bf1aff0f1aca

With this, my reproducer above doesn't happen (no stuck nodes).

Getting closer...

**Comment by colinsullivan** [comment]
> What does Link start/stop mean for TempoClocks? We don't have that concept.

FWIW - I'd make use of `LinkClock` start / stop callbacks to stop and start `EventStreamPlayer` instances or start synths.

```supercollider
l = LinkClock(1);
p = Pbind(\instrument, \click, \freq, Pseq([900, Pn(400,3)],inf));

l.onLinkStart({
	p.play(l, quant:4);

	// synth gate on, etc.
});
l.onLinkStop({
	p.stop();

	// synth gate off, etc.
});

```

**Comment by jamshark70** [comment]
> FWIW - I'd make use of `LinkClock` start / stop callbacks to stop and start `EventStreamPlayer` instances or start synths.

I tested this. That's actually feasible. Link transmits the start/stop request at the moment it happens in the other peer. If the user-supplied callback uses `quant` appropriately (as in your hypothetical example), then it should work.

There is something funny that happens if we start our `LinkClock` a long time after the sclang process starts. While I was testing against QLinkHut, I found that a late-started LinkClock is out of sync with QLinkHut. I think this is because we set our `linkTimeOfInitialization` when sclang starts, and then measure the system clock against this time, while QLinkHut grabs its init time and then measures the sample clock against it.

I found, in a 30-minute-plus test, that if LinkClock and QLinkHut start in sync, then they stay in sync -- the network messaging compensates for system-vs-sample clock drift. But, if we don't activate the messaging right away, then we can't be sure of being in sync with apps that are based on the sample clock (and we can't expect DAWs *not* to follow the sample clock).

Still trying to figure out the conditions when sync fails. It may be: sclang and QLinkHut are started at very different times, or sclang's LinkClock is started long after the process started (and the other process has been running on the sample clock for a while), or...? I was trying to break it just now and couldn't.

**Comment by jpburstrom** [comment]
FWIW I've just tried https://github.com/jamshark70/supercollider/commit/45054f885608424092606edb0f820c8fa5d6a854  with Live 9.7.7 on MacOS, setting both LinkClock and server latency to 0.2, and everything seems to work quite nicely! I tried a few extreme tempo changes, and it seemed to handle it very well. When recording simultaneous clicks from Live and SC (through loopback on my sound card), while starting/stopping and changing tempo, it seems the two can get a bit randomly off sync (most of the time around 32 samples), but maybe that's due to the loopback processing, I don't know.

**Comment by jamshark70** [comment]
It will never be 100% perfect because it depends on network transport, where timing is unreliable. (Just wait until you try it with 2 machines -- network timing jitter may be as wide as 200 ms!) If it drifts out of sync by a few ms and then locks back in, that's as good as it's going to get.

I opened a github issue with Ableton about tempo changes, and they say a brief moment of drift is expected when the tempo changes.

It sounds like a successful test, then, thank you for trying it!

**Comment by mzero** [comment]
Haven't forgotten this... just had two performances this past week... and
another this coming Tuesday. But, I'll try to get to trying this out today
or tomorrow or Wednesday at latest.

BTW: All those performances were/are using the version pre the link latency
patch - and Wednesday's used it with other users on Ableton Live via Link
(over WiFi). Worked well, and the latenncy (50 ~ 100ms I think, by ear) was
tolerable for the kind of thing we were sync'ing - though would clearly be
too much for percussive work.

**Comment by jamshark70** [comment]
I'm tidying up the last few things, and the help, and I have a question for other interested parties.

I would like to remove the `tempoChanged` callback function.

TempoClock has, for years, broadcasted a `\tempo` notification whenever its tempo changes. Originally, LinkClock did not do this, but I fixed that.

This makes it redundant to have a separate `tempoChanged` function.

I suggest interface parity between the two classes. For TempoClock, if some client object needs to know about tempo changes, it registers a dependent. LinkClock should not be any different. Also, it's a bit silly for LinkClock to provide a second way that is not duplicated in TempoClock.

Note also that the `beats` and `secs` values passed into the current `tempoChanged` function are based on the time when the tempo change was received on the local machine, which is guaranteed not to be the exact moment when the tempo change was sent (it may be delayed by a few hundred ms -- but, the *sent* time reflects the musical intention and is more important). I don't see the use of passing this information into a user function.

But, maybe somebody really, really likes the current callback and can argue against removing it? I don't want to take it out unilaterally.

**Comment by mzero** [comment]
tl;dr: It works!

I pulled `topic/hjhLinkLatency` and tried it today:

* rebased onto Version_3_10_1
* built on RasberryPi under Raspian Stretch Lite, 2018-11-13

My test was:

* `LinkClock` on RPi, sync'ing via Link with iMac, both over ethernet
* a routine running every beat, playing a simple `SynthDef`
* `scsynth` is playing, via Jack to a Pisound card
* Pisound card audio out, connected to audio interface in on iMac (AudioHub)
* Live 10.0.6 records this, and is sync'd via Link.

At 120 bpm, there was 20ms of latency. Setting `metro.latency = 0.020;` worked perfectly. Everything is spot on.

Changing the tempo (both via Live, and via my script) keeps things almost perfectly aligned. I notice that at 80 bpm it is about 1 ms too much latency, and at 180 bpm it is 1 ms too short. 

My sc script is far from minimal. It is my normal performance script and the full set up is described here: [Live w/Digitakt & Pisound](https://www.elektronauts.com/t/live-w-digitakt-pisound-w-video-visuals/61079/3).

When I set this up for performance - where the sc script generates MIDI clock, sent over USB to the drum machine, who's audio is then piped into the Pisound card and scsynth for processing, and finally back out again.... The whole thing needs 44ms of latency at 120bpm. Saddly, in this configuration, there is much more variability in the latency at different tempos. Not sure, but I'm suspecting my script (which is doing quite a lot at the MIDI clock level.)

**There is one outstanding issue:** The `\tempo` changed notification generated from `prTempoChanged` doesn't seem to be working. I still had to set a `tempoChanged` handler to get tempo updates.  I'll investigate this more.

Finally - I think the way to think about future scheduled tempo changes in Link is that "future" is a local concept. If P2 schedules a tempo change in the future, in the future it is like P2 "twists the tempo knob" at that point - which means that other Link users won't see the tempo change at the exact same point in beat time..... But, they will (eventually) be at the same tempo, and still have the beats aligned. Just during the change things might be slightly out. I think this is okay - in practice, even for non-dance music (my group often plays, link enabled, down at 80bpm or less... then speeding up to 130 or more) this will be fine. Live performed tempo changes are always a bit of fuzz.

**Comment by mzero** [comment]
Also - meant to add I'm totally okay with removing the `tempoChanged` callback.

**Comment by mzero** [comment]
Thought: perhaps we need a `changed(\numPeers)` notification? Link gives a callback for seeing when that changes. Since it is really nice to display that value to the user, it would be nice not to have to poll for it.

**Comment by jamshark70** [comment]
Great! Another good test.

A `numPeers` notification is a good idea, and should be easy enough, following the existing pattern. I can do it later.

**Comment by colinsullivan** [comment]
I have successfully tested this as well, using the `topic/hjhLinkLatency` branch.  I am able to sync with Ableton and record from SuperCollider into Ableton using Jack on macOS 10.14.

I first leave the latency unset, then record in Ableton, then measure offset (roughly).  Setting the latency to `0.09` seems to get it close enough (it over-compensates, then I can use the track delay in Ableton if I want it sample-accurate).

Moving the tempo up slowly to 400BPM, everything kept in sync.  Then I abruptly set it to 50, then to 120...It all looked / sounded good to me.

I've been using the `topic/ableton-link` branch extensively for a few months now.  The latency fixes make it all the better.

I would _love_ those start / stop hooks - looks like the structure _could_ be pretty similar to what is happening for the tempo callback?  The Link API call is [here](https://github.com/Ableton/link/blob/c1449edb28ccba86396a46a9989b8795960160a9/include/ableton/Link.hpp#L149).  That said, I'm happy to call it feature creep and advocate for it later...

**Comment by jamshark70** [comment]
> I first leave the latency unset, then record in Ableton, then measure offset (roughly). Setting the latency to `0.09` seems to get it close enough (it over-compensates, then I can use the track delay in Ableton if I want it sample-accurate).

I'd like to suggest that this is not the ideal way to use, and to test, Link support.

*All* audio apps have to account for latency. Usually this is audio driver latency. If I'm reading Ableton's documentation correctly, Link determines that, say, beat 176.0 corresponds to 12:31:57.37463 on the local system clock, and it's the Link-enabled application's responsibility to make sure that audio belonging to beat 176.0 hits the speakers at exactly 12:31:57.37463.

In SC, scsynth is supposed to handle audio driver latency by itself -- so, if it receives a message with a timestamp for 12:31:57.37463, then this is when the sound should hit the speakers. So we need to be sure that sclang's timestamp corresponding to beat 176.0 is 12:31:57.37463.

The timestamp is generated by adding `latency` to the current SystemClock logical time. If server messaging latency is 0.2 sec, then a message that should sound at SystemClock logical time 12:31:57.37463 needs to be sent at 12:31:57.37463 - 0.2 = 12:31:57.17463. `LinkClock.latency = 0.2` is supposed to take care of that (our beats-->seconds call, for LinkClock, does the equivalent of "beats2secs(beats) - latency").

Ideally, then: you set the clock's latency to match server messaging latency. That *should* be it. If *many* users find that this is consistently too early, or too late, then there may be something wrong in the way SC is handling the latency, and we need to fix it, not work around it. But we can never figure that out, if everybody takes the approach of "I'm pretty sure it isn't going to work, so I'm going to measure how it isn't working and then manually adjust."

TL;DR - just set the latency. How does it work in that case?

> I would _love_ those start / stop hooks...

If you're using an up-to-date branch, then you already have the hooks.

**Comment by mzero** [comment]
I think in @colinsullivan 's case, there is extra latency from the routing of SC via jack back into Ableton. You can either compensate for that in Ableton (treating jack as an audio interface that delivers audio with 90ms of delay), or you can compensate for that in `LinkClock`. The advantage of the later is that while actually monitoring the recording, everything will sound in sync. If you are performing live, this is what you want. Post recording, on playback, either style of compensation will work.

**Comment by jamshark70** [comment]
@mzero thanks for explaining.

That's relevant, then, for the discussions about unit testing. If we need to record scsynth and linkhut into the same file, and the loopback itself incurs latency, it's messy (unless you have one scsynth, and linkhut, both feeding into a second scsynth that is recording).

**Comment by jamshark70** [comment]
Superseded by #4331. I believe this can be closed now.
